ContactDetailsApi/V2/UseCase/FetchAllContactDetailsByUprnUseCase.cs

```diff
@@ -5,24 +5,100 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
+using ContactDetailsApi.V1.Boundary.Request;
+using ContactDetailsApi.V2.Domain;
+using Hackney.Shared.Person.Infrastructure;
+using Hackney.Shared.Tenure.Infrastructure;
+using Hackney.Core.Logging;
 
 namespace ContactDetailsApi.V2.UseCase
 {
     public class FetchAllContactDetailsByUprnUseCase : IFetchAllContactDetailsByUprnUseCase
     {
 
-        private readonly IContactDetailsGateway _gateway;
+        private readonly ITenureDbGateway _tenureGateway;
+        private readonly IPersonDbGateway _personGateway;
+        private readonly IContactDetailsGateway _contactGateway;
 
-        public FetchAllContactDetailsByUprnUseCase(IContactDetailsGateway gateway)
+        public FetchAllContactDetailsByUprnUseCase(ITenureDbGateway tenureGateway, IPersonDbGateway personGateway, IContactDetailsGateway contactGateway)
         {
-            _gateway = gateway;
+            _tenureGateway = tenureGateway;
+            _personGateway = personGateway;
+            _contactGateway = contactGateway;
         }
 
+        [LogCall]
         public async Task<List<ContactByUprn>> ExecuteAsync()
         {
-            var results = await _gateway.FetchAllContactDetailsByUprnUseCase();
+            var tenures = await _tenureGateway.GetAllTenures().ConfigureAwait(false);
+            tenures = tenures.Where(x => x.TenuredAsset?.Uprn != null).ToList();
 
-            return results;
+            var personIds = tenures.Select(x => x.HouseholdMembers.Select(y => y.Id))
+                                   .SelectMany(x => x)
+                                   .Distinct()
+                                   .ToList();
+
+            var personsData = await _personGateway.GetPersons(personIds).ConfigureAwait(false);
+            var persons = personsData.ToDictionary(x => x.Id, x => x);
+
+            var contactDetails = new Dictionary<Guid, List<ContactDetails>>();
+            foreach (var personId in personIds)
+            {
+                var data = await _contactGateway
+                    .GetContactDetailsByTargetId(new ContactQueryParameter { TargetId = personId })
+                    .ConfigureAwait(false);
+                contactDetails.Add(personId, data);
+            }
+            return ConsolidateData(tenures, persons, contactDetails);
```

Comment:
Two things. 

1. I think this would be cleaner if we added a `GetContactDetails` method.
2. Are we fetching a lot of contact details? It could be fairly slow running these in series. Might it be worth batch getting them. For example [https://www.michalbialecki.com/2018/04/19/how-to-send-many-requests-in-parallel-in-asp-net-core/](https://www.michalbialecki.com/2018/04/19/how-to-send-many-requests-in-parallel-in-asp-net-core/)

Reply:
Agree with @LBHCallumM on bullet #1 here. Splitting out the code into Private encapsulated methods  makes the code much more readable because we know the intent of the code. Also makes the code more maintainable in the long term